File: elixir/say/lib/say.ex
defmodule Say do
  @words %{
    0 => "zero",
    1 => "one",
    2 => "two",
    3 => "three",
    4 => "four",
    5 => "five",
    6 => "six",
    7 => "seven",
    8 => "eight",
    9 => "nine",
    10 => "ten",
    11 => "eleven",
    12 => "twelve",
    13 => "thirteen",
    14 => "fourteen",
    15 => "fifteen",
    16 => "sixteen",
    17 => "seventeen",
    18 => "eighteen",
    19 => "nineteen",
    20 => "twenty",
    30 => "thirty",
    40 => "forty",
    50 => "fifty",
    60 => "sixty",
    70 => "seventy",
    80 => "eighty",
    90 => "ninety"
  }

  @billions 1_000_000_000
  @millions 1_000_000
  @thousands 1_000
  @hundreds 100

  defguardp is_even_ten(number)
            when is_integer(number) and number in 10..90 and rem(number, 10) == 0

  defguardp is_even_hundred(number)
            when is_integer(number) and number in @hundreds..900 and rem(number, @hundreds) == 0

  defguardp is_even_million(number)
            when is_integer(number) and number in @millions..900_000_000 and
                   rem(number, @millions) == 0

  defguardp is_even_billion(number)
            when is_integer(number) and number in @billions..900_000_000_000 and
                   rem(number, @billions) == 0

  defguardp is_even_thousand(number)
            when is_integer(number) and number in @thousands..9000 and
                   rem(number, @thousands) == 0

  @doc """
  Translate a positive integer into English.
  """
  @spec in_english(integer) :: {atom, String.t()}

  def in_english(number) when number not in 0..999_999_999_999 do
    {:error, "number is out of range"}
  end

  def in_english(number), do: {:ok, say(number)}

  defp number_of_unit(number, unit), do: trunc(number / unit)

  defp lookup(number), do: Map.get(@words, number, number)

  defp say(number) when is_even_billion(number) do
    billions = number_of_unit(number, @billions)

    "#{say(billions)} billion"
  end

  defp say(number) when number >= @billions do
    billions = number_of_unit(number, @billions)

    "#{say(billions)} billion #{say(number - billions * @billions)}"
  end

  defp say(number) when is_even_million(number) do
    millions = number_of_unit(number, @millions)

    "#{say(millions)} million"
  end

  defp say(number) when number >= @millions do
    millions = number_of_unit(number, @millions)

    "#{say(millions)} million #{say(number - millions * @millions)}"
  end

  defp say(number) when number > @thousands do
    thousands = number_of_unit(number, @thousands)

    "#{say(thousands)} thousand #{say(number - thousands * @thousands)}"
  end

  defp say(number) when is_even_thousand(number) do
    thousands = number_of_unit(number, @thousands)

    "#{say(thousands)} thousand"
  end

  defp say(number) when is_even_hundred(number) do
    hundreds = number_of_unit(number, @hundreds)

    "#{say(hundreds)} hundred"
  end

  defp say(number) when number > @hundreds do
    hundreds = number_of_unit(number, @hundreds)

    "#{say(hundreds)} hundred #{say(number - hundreds * @hundreds)}"
  end

  defp say(number) when number in 0..19, do: lookup(number)
  defp say(number) when is_even_ten(number), do: lookup(number)

  defp say(number) when number > 20 do
    tens = number_of_unit(number, 10)

    "#{say(tens * 10)}-#{say(number - tens * 10)}"
  end
end

File: clojure/say/src/say.clj
(ns say
  (:require [clojure.spec.alpha :as alpha]
            [clojure.string :as str]))

(def num-words ["zero",
                "one",
                "two",
                "three",
                "four",
                "five",
                "six",
                "seven",
                "eight",
                "nine",
                "ten",
                "eleven",
                "twelve",
                "thirteen",
                "fourteen",
                "fifteen",
                "sixteen",
                "seventeen",
                "eighteen",
                "nineteen",
                "twenty"])

(def tens-words ["",
                 "",
                 "twenty",
                 "thirty",
                 "forty",
                 "fifty",
                 "sixty",
                 "seventy",
                 "eighty",
                 "ninety"])

(defn- is-num-valid? [num] (alpha/int-in-range? 0 1000000000000 num))

(defn- find-word-for-digits [num] (nth num-words num))
(defn- find-word-for-tens [num] (nth tens-words num))

(defn- clean [word]
  (-> word
      (str/replace #"\-$" "")
      (str/trim)))

(defn- str-to-int
  ([str] (Integer/parseInt str))
  ([str start end] (str-to-int (subs str start end))))

(defn- zero-pad-to-length [num length] (format (str "%0" length "d") num))

(defn- assemble-three-digit-number [num hundreds tens digits]
  (str
   (when (> num 99) (str (find-word-for-digits hundreds) " hundred "))
   (when (> num 9) (str (find-word-for-tens tens) "-"))
   (find-word-for-digits digits)))

(defn- format-number
  ([num] (format-number num ""))
  ([num unit]
   (let [digits (str-to-int num)
         full-num (zero-pad-to-length digits 3)
         formatted-number (assemble-three-digit-number
                           digits
                           (str-to-int full-num 0 1)
                           (str-to-int full-num 1 2)
                           (str-to-int full-num 2 3))]
     (if (= "zero" formatted-number)
       ""
       (format "%s %s "
               formatted-number
               unit)))))

(defn- word-for-num [num]
  (let [full-num (zero-pad-to-length num 12)]
    (if (<= num (count num-words))
      (find-word-for-digits num)
      (clean (str
              (format-number (subs full-num 0 3) "billion")
              (format-number (subs full-num 3 6) "million")
              (format-number (subs full-num 6 9) "thousand")
              (format-number (subs full-num 9 10) "hundred")
              (str (find-word-for-tens (str-to-int full-num 10 11)) "-")
              (format-number (subs full-num 11 12)))))))

(defn number [num]
  (if (is-num-valid? num)
    (word-for-num num)
    (throw (IllegalArgumentException.))))
